libvast/src/evaluate.cpp

```diff
@@ -0,0 +1,465 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2022 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "vast/arrow_table_slice.hpp"
+#include "vast/bitmap_algorithms.hpp"
+#include "vast/detail/assert.hpp"
+#include "vast/detail/overload.hpp"
+#include "vast/detail/passthrough.hpp"
+#include "vast/expression.hpp"
+#include "vast/ids.hpp"
+#include "vast/logger.hpp"
+#include "vast/table_slice.hpp"
+#include "vast/type.hpp"
+
+#include <arrow/record_batch.h>
+
+#include <cstddef>
+#include <span>
+
+namespace vast {
+
+namespace {
+
+// A utility function for evaluating meta extractors in predicates. This is
+// always a yes or no question per batch, so the function does not have to deal
+// with bitmaps at all.
+bool evaluate_meta_extractor(const table_slice& slice,
+                             const meta_extractor& lhs, relational_operator op,
+                             const data& rhs) {
+  switch (lhs.kind) {
+    case meta_extractor::kind::type:
+      return evaluate(materialize(slice.layout().name()), op, rhs);
+    case meta_extractor::kind::field: {
+      const auto* s = caf::get_if<std::string>(&rhs);
+      if (!s) {
+        VAST_WARN("#field can only compare with string");
+        return false;
+      }
+      auto result = false;
+      auto neg = is_negated(op);
+      for (const auto& layout_rt = caf::get<record_type>(slice.layout());
+           const auto& [field, index] : layout_rt.leaves()) {
+        const auto fqn
+          = fmt::format("{}.{}", slice.layout().name(), layout_rt.key(index));
+        // This is essentially s->ends_with(fqn), except that it also checks
+        // the dot separators correctly (modulo quoting).
+        const auto [fqn_mismatch, s_mismatch]
+          = std::mismatch(fqn.rbegin(), fqn.rend(), s->rbegin(), s->rend());
+        if (s_mismatch == s->rend()
+            && (fqn_mismatch == fqn.rend() || *fqn_mismatch == '.')) {
+          result = true;
+          break;
+        }
+      }
+      return neg != result;
+    }
+    case meta_extractor::kind::import_time:
+      return evaluate(data{slice.import_time()}, op, rhs);
+  }
+  __builtin_unreachable();
+}
+
+template <relational_operator Op>
+struct cell_evaluator;
```

Comment:
NB: I really like the way you simplified the problem from triple dispatch to double dispatch!

Reply:
Thanks!